packages/core/injector/compiler.ts

```diff
@@ -12,36 +12,58 @@ export interface ModuleFactory {
 }
 
 export class ModuleCompiler {
-  constructor(private readonly moduleTokenFactory = new ModuleTokenFactory()) {}
+  constructor(
+    private readonly _moduleOpaqueKeyFactory: ModuleOpaqueKeyFactory,
+  ) {}
+
+  get moduleOpaqueKeyFactory(): ModuleOpaqueKeyFactory {
+    return this._moduleOpaqueKeyFactory;
+  }
 
   public async compile(
-    metatype: Type<any> | DynamicModule | Promise<DynamicModule>,
+    moduleClsOrDynamic:
+      | Type
+      | DynamicModule
+      | ForwardReference
+      | Promise<DynamicModule>,
   ): Promise<ModuleFactory> {
-    const { type, dynamicMetadata } = this.extractMetadata(await metatype);
-    const token = this.moduleTokenFactory.create(type, dynamicMetadata);
+    moduleClsOrDynamic = await moduleClsOrDynamic;
+
+    const { type, dynamicMetadata } = this.extractMetadata(moduleClsOrDynamic);
+    const token = dynamicMetadata
+      ? this._moduleOpaqueKeyFactory.createForDynamic(
+          type,
+          dynamicMetadata,
+          moduleClsOrDynamic as DynamicModule | ForwardReference,
+        )
+      : this._moduleOpaqueKeyFactory.createForStatic(
+          type,
+          moduleClsOrDynamic as Type,
+        );
+
     return { type, dynamicMetadata, token };
   }
 
   public extractMetadata(
-    metatype: Type<any> | ForwardReference | DynamicModule,
+    moduleClsOrDynamic: Type | ForwardReference | DynamicModule,
   ): {
-    type: Type<any>;
+    type: Type;
     dynamicMetadata?: Partial<DynamicModule> | undefined;
   } {
-    if (!this.isDynamicModule(metatype)) {
+    if (!this.isDynamicModule(moduleClsOrDynamic)) {
       return {
-        type: (metatype as ForwardReference)?.forwardRef
-          ? (metatype as ForwardReference).forwardRef()
-          : metatype,
+        type: (moduleClsOrDynamic as ForwardReference)?.forwardRef
+          ? (moduleClsOrDynamic as ForwardReference).forwardRef()
+          : moduleClsOrDynamic,
       };
```

Comment:
Just keep the same structure for hidden classes.

```suggestion
          : moduleClsOrDynamic,
          dynamicMetadata: undefined,
      };
```